### DISCUSSION TITLE
Getting "rate limited" trying to authenticate for the first time

### DISCUSSION THREAD

**Discussion by krisalyssa** [bluesky-social/atproto#2971] [How to do it?] [answered]
Following the instructions for creating a session via `curl` at https://docs.bsky.app/docs/get-started :

```shell
% BLUESKY_HANDLE="krisalyssa.bsky.social" BLUESKY_PASSWORD="****-****-****-****" \
curl -X POST https://bsky.social/xrpc/com.atproto.server.createSession \
-H "Content-Type: application/json" \
-d '{"identifier": "'"$BLUESKY_HANDLE"'", "password": "'"$BLUESKY_PASSWORD"'"}'
```

where `BLUESKY_PASSWORD` is an app password I created in my account, and I get `{"error":"RateLimitExceeded","message":"Rate Limit Exceeded"}`.

Since this is the first time I've tried this, I'm reasonably confident that I'm not actually exceeding a rate limit.

**Comment by fxcoudert** [comment]
Same here. I have connected to the account only a few times, and now:

```
>>> bluesky.login('xxxxxxxxx.bsky.social', 'xxxxxxxxxx')
```

gives

```
atproto_client.exceptions.RequestException: Response(success=False, status_code=429, content=XrpcError(error='RateLimitExceeded', message='Rate Limit Exceeded'), headers={'date': 'Tue, 19 Nov 2024 11:41:04 GMT', 'content-type': 'application/json; charset=utf-8', 'content-length': '61', 'connection': 'keep-alive', 'x-powered-by': 'Express', 'access-control-allow-origin': '*', 'ratelimit-limit': '100', 'ratelimit-remaining': '0', 'ratelimit-reset': '1732088805', 'ratelimit-policy': '100;w=86400', 'etag': 'W/"3d-egyFOcXCrKQFzoI7522/4+PpxIk"', 'vary': 'Accept-Encoding'})
```

I definitely have not connected 100 times.

**Comment by pbitkowski** [comment]
i have the same problem

**Comment by dom96** [comment]
Does the limit reset if you wait a day? It seems to for me.

**Comment by krisalyssa** [comment]
@dom96 It's failing on the first attempt. So, no.

**Comment by krisalyssa** [comment]
Okay, I was able to get authentication and posting via `curl` to work today. I hardcoded the values for `BLUESKY_HANDLE` and `BLUESKY_PASSWORD` instead of interpolating environment variables into the body.

I'll try again with something closer to what the documentation has. After that, I'll follow up with nicely-formatted commands to demonstrate what worked for me.

**Reply by krisalyssa** [reply, marked as answer]
I'm sort of ashamed to admit this, but ultimately the problem was that I was setting the environment variables as a prefix to the `curl` command rather than explicitly exporting them prior.

This works:

```shell
export BLUESKY_HANDLE="..."
export BLUESKY_PASSWORD="..."
export BLUESKY_PDS_HOST="https://bsky.social"
curl --request POST --header "Content-type: application/json" --location --url "https://bsky.social/xrpc/com.atproto.server.createSession" --data-raw "{\"identifier\":\"${BLUESKY_HANDLE}\",\"password\":\"${BLUESKY_PASSWORD}\"}"                                                                                                                   
```

where `BLUESKY_HANDLE` is the full handle (e.g. `foo.bsky.social`) and `BLUESKY_PASSWORD` is an app password created in the account named by `BLUESKY_HANDLE`.
